scripts/variantstore/wdl/extract/import_gvs.py

```diff
@@ -397,9 +393,12 @@ def convert_array_with_id_keys_to_dense_array(arr, ids, drop=[]):
 
         vd = vd.drop('allele_NO', 'allele_YES', 'allele_is_snp', 'allele_OK')
 
-        if unphase:
-            vd = vd.annotate_entries(LGT=vd.LGT.unphase())
-        vd = vd.select_entries(GT=hl.vds.lgt_to_gt(vd.LGT, vd.LA), **vd.entry)
+        # Clean up any completely empty phasing fields.
+        e = vd.entries()
+        if e.all(hl.is_missing(e.PS)):
+            vd = vd.drop('PS')
+        if e.all(hl.is_missing(e.PID)):
+            vd = vd.drop('PID')
```

Comment:
`MatrixTable.entries()` converts the compact matrix table representation:
```
row_key col1 col2 col3 ...
1 a b c ...
2 d e f ...
```

into a "coordinate-wise" `Table` representation which is not compact:
```
row_key col_key entry_struct
1 col1 a
1 col2 b
1 col3 c
...
2 col1 d
...
```

Notice that this duplicates the row and column keys $O(N_{entries})$ times and splits the compact arrays of entry values into individual elements. There are a few issues with this:

1. Storing an entries table on disk is substantially larger than storing a `MatrixTable` b/c of the repeated keys. [1]
2. Because of (2) `hl.vds.VariantDataset` requires the variant_data to be a `MatrixTable`, not a `Table`.
3. Hail amortizes the cost of decoding and of memory allocations [2] per-row. The `Table` representation introduces $N_{samples}$ allocations (and, if it were written, would introduce $N_{samples}$ encodes and, on a subsequent read, decodes).

Instead, I recommend staying in the matrix representation. `Table.all` is literally defined as:
```python3
    ...
    return self.aggregate(hl.agg.all(expr))
```
The equivalent, for entry-fields, in `MatrixTable` is:
```python3
if vd.aggregate_entries(hl.agg.all(hl.is_missing(vd.PS)):
    vd = vd.drop('PS')
if vd.aggregate_entries(hl.agg.all(hl.is_missing(vd.PID)):
    vd = vd.drop('PID')
```
However, I wouldn't stop here. Each call to aggregate is an *action* which executes the entire Hail pipeline up to this point. Hail has a number of optimization passes that should be able to eliminate all the intermediate manipulations from lines 315 to 394 and execute instead this:
```python3
vds = hl.vds.read_vds(...)
vds.vd.aggregate_entries(hl.agg.all(hl.is_missing(vds.vd.PS)))
```
But even still, because we use `aggregate_entries` twice, we must perform the distributed execution twice. This execution is dominated by reading and decoding time (see [1] again). Instead of scanning the data twice, we can scan it once:
```python3
ps_all_missing, pid_all_missing = vd.aggregate_entries((
    hl.agg.all(hl.is_missing(vd.PS)),
    hl.agg.all(hl.is_missing(vd.PID))
)
if ps_all_missing:
    vd = vd.drop('PS')
if pid_all_missing:
    vd = vd.drop('PID')
```


[1] We're actively and eagerly pursuing a column-oriented storage format that would reduce this overhead, but, alas, we do not have that today.
[2] We do something called "region" or "arena" memory management [Wikipedia](https://en.wikipedia.org/wiki/Region-based_memory_management). Allocation is cheap (increment a pointer) and deallocating all allocations in the region (aka the memory used to process one row) just resets the pointer to zero.

Reply:
Thank you for the detailed explanation Dan, I replaced my original code with this much faster, better version! 🚀